sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,718 @@
+                {
+                    "auto",
+                    "brute",
+                    "kdtree",
+                    "balltree",
+                }
+            )
+        ],
+        "leaf_size": [Interval(Integral, left=1, right=None, closed="left")],
+        "memory": [str, None, Path],
+        "n_jobs": [Integral, None],
+        "cluster_selection_method": [StrOptions({"eom", "leaf"})],
+        "allow_single_cluster": ["boolean"],
+        "metric_params": [dict, None],
+    }
+
+    def __init__(
+        self,
+        min_cluster_size=5,
+        min_samples=None,
+        cluster_selection_epsilon=0.0,
+        max_cluster_size=0,
+        metric="euclidean",
+        alpha=1.0,
+        algorithm="auto",
+        leaf_size=40,
+        memory=None,
+        n_jobs=4,
+        cluster_selection_method="eom",
+        allow_single_cluster=False,
+        metric_params=None,
+    ):
+        self.min_cluster_size = min_cluster_size
+        self.min_samples = min_samples
+        self.alpha = alpha
+        self.max_cluster_size = max_cluster_size
+        self.cluster_selection_epsilon = cluster_selection_epsilon
+        self.metric = metric
+        self.algorithm = algorithm
+        self.leaf_size = leaf_size
+        self.memory = memory
+        self.n_jobs = n_jobs
+        self.cluster_selection_method = cluster_selection_method
+        self.allow_single_cluster = allow_single_cluster
+        self.metric_params = metric_params
+
+    def fit(self, X, y=None):
+        """Perform HDBSCAN clustering from features or distance matrix.
+
+        Parameters
+        ----------
+        X : array or sparse (CSR) matrix of shape (n_samples, n_features), or \
+                array of shape (n_samples, n_samples)
+            A feature array, or array of distances between samples if
+            `metric='precomputed'`.
+
+        y : Ignored
+            Ignored.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        self._validate_params()
+        self._metric_params = self.metric_params or {}
+        if self.metric != "precomputed":
+            # Non-precomputed matrices may contain non-finite values.
+            X = self._validate_data(
+                X, accept_sparse="csr", force_all_finite=False, dtype=np.float64
+            )
+            self._raw_data = X
+
+            all_finite = (
+                np.all(np.isfinite(X.data)) if issparse(X) else np.all(np.isfinite(X))
+            )
+
+            if not all_finite:
+                # Pass only the purely finite indices into hdbscan
+                # We will later assign all non-finite points to the
+                # background-1 cluster
+                finite_index = get_finite_row_indices(X)
+                X = X[finite_index]
+                internal_to_raw = {x: y for x, y in enumerate(finite_index)}
+                outliers = list(set(range(X.shape[0])) - set(finite_index))
```

Comment:
Going down this route, I've allocated `-2` as a special label to denote _missing_ data, keeping `-1` as the label for _outlier_ data. Similarly, while outlier data has probability `0`, missing data has probability `np.nan`.

Alternatives considered:
- Return the final `labels_` as a NumPy masked array, which are not well-supported right now. Plus I would bet most users won't have much experience dealing with them.
- Shift labels by one so that outliers are label `0` and missing data are label `-1`. While I think this is sensible, I don't want to diverge from the original HDBSCAN label semantics too much, and it would still be surprising to some users who are used to `0`-indexing.